php-checks/src/main/java/org/sonar/php/checks/utils/AbstractDuplicateBranchCheck.java

```diff
@@ -19,96 +19,57 @@
+import org.sonar.plugins.php.api.tree.statement.StatementTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
 
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Set;
 
-public abstract class AbstractDuplicateBranchCheck extends SquidCheck<LexerlessGrammar> {
+public abstract class AbstractDuplicateBranchCheck extends PHPVisitorCheck {
 
-  private final Set<AstNode> checkedIfStatements = Sets.newHashSet();
-
-  protected abstract AstNodeType ifBranchNodeType();
-
-  protected abstract AstNodeType caseClauseChildType();
-
-  protected abstract void addIssue(String type, AstNode duplicate, AstNode duplicated);
+  protected List<IfStatementTree> checkedIfStatements;
 
   @Override
-  public void leaveFile(AstNode astNode) {
-    checkedIfStatements.clear();
+  public void visitScript(ScriptTree tree) {
+    checkedIfStatements = new ArrayList<>();
+    super.visitScript(tree);
   }
 
-  @Override
-  public void init() {
-    subscribeTo(PHPGrammar.IF_STATEMENT, PHPGrammar.CASE_LIST);
-  }
+  protected abstract void raiseIssue(String branchType, Tree duplicatedTree, Tree duplicatingTree);
 
-  @Override
-  public void visitNode(AstNode node) {
-    if (node.is(PHPGrammar.IF_STATEMENT)) {
-      visitIfStatementNode(node);
-    } else {
-      visitCaseListNode(node);
-    }
-  }
+  protected List<Tree> getClauses(IfStatementTree ifStatement) {
+    Preconditions.checkArgument(ifStatement.is(Kind.IF_STATEMENT));
 
-  private void visitIfStatementNode(AstNode node) {
-    if (checkedIfStatements.contains(node)) {
-      return;
-    }
-    List<AstNode> branchNodes = Lists.newArrayList();
-    AstNode ifNode = node;
-    while (ifNode != null) {
-      checkedIfStatements.add(ifNode);
-      branchNodes.add(ifNode.getFirstChild(ifBranchNodeType()));
-      AstSelect elseIfNodes = ifNode.select()
-        .children(PHPGrammar.ELSEIF_LIST)
-        .children(PHPGrammar.ELSEIF_CLAUSE)
-        .children(ifBranchNodeType());
-      for (AstNode condition : elseIfNodes) {
-        branchNodes.add(condition);
-      }
-      AstSelect elseNodes = ifNode.select()
-        .children(PHPGrammar.ELSE_CLAUSE);
-      for (AstNode elseNode : elseNodes.children(ifBranchNodeType())) {
-        branchNodes.add(elseNode);
-      }
-      AstSelect elseIfs = elseNodes
-        .children(PHPGrammar.STATEMENT)
-        .children(PHPGrammar.IF_STATEMENT);
-      ifNode = elseIfs.isEmpty() ? null : elseIfs.get(0);
-    }
-    checkBranches(branchNodes, "branch");
-  }
+    List<Tree> clauses = new ArrayList<>();
+    clauses.add(ifStatement);
+    clauses.addAll(ifStatement.elseifClauses());
 
-  private void visitCaseListNode(AstNode node) {
-    AstSelect conditions = node.select()
-      .children(PHPGrammar.CASE_CLAUSE)
-      .children(caseClauseChildType());
-    checkBranches(ImmutableList.copyOf(conditions), "case");
-  }
+    if (ifStatement.elseClause() != null) {
```

Comment:
@vilchik-elena I think that this condition is redundant with the next `while`statement.